### PR TITLE
Update TTS CI

### PR DESCRIPTION
**Pull request by blisc**
Can be merged after the current release.

TTS Tests Updates
- ~Update step TTS Tests to fix random seed~
- ~Update step TTS Tests to check loss on entire evaluation dataset rather than last train batch loss~
- ~Update step TTS Tests to do teacher forcing during eval~
- Fix test_tacotron2_training test and test_waveglow_training
- Disable Tacotron2 test during Jenkins in favour of the integration test

General Updates
- Add an EvaluatorCallback option that stops doing eval on step 0
- Change the final message that logs training script duration to use datetime
- Add a pytest marker `skipduringci` to skip some pytest tests during the CI pipeline.

### CODE REVIEW

File: examples/asr/jasper_an4.py [hunk 1]
```diff
@@ -95,6 +95,7 @@ def create_dags(model_config_file, vocab, args, nf):
         user_epochs_done_callback=process_evaluation_epoch,
         eval_step=args.eval_freq,
         tb_writer=nf.tb_writer,
+        eval_at_start=not args.do_not_eval_at_start,
```
Comment (tkornuta-nvidia) [inline examples/asr/jasper_an4.py]: I would suggest going even further and option to turn evaluation off at all (during testing we do not check this value at all, am I right?)

File: examples/tts/tacotron2.py [hunk 2]
```diff
@@ -245,6 +253,7 @@ def create_eval_dags(
             tb_writer_func=partial(tacotron2_eval_log_to_tb_func, tag=tagname),
             eval_step=eval_freq,
             tb_writer=neural_factory.tb_writer,
+            eval_at_start=not do_not_eval_at_start,
```
Comment (tkornuta-nvidia) [inline examples/tts/tacotron2.py]: ehehehe 👍

File: nemo/core/neural_factory.py [hunk 3]
```diff
@@ -339,6 +339,11 @@ def __init__(
                 np.random.seed(random_seed)
                 random.seed(random_seed)
```
Comment (tkornuta-nvidia) [inline nemo/core/neural_factory.py]: 👍 I had something similar in PTP, additionally I was logging the seed even if the user has not set it manually:

https://github.com/IBM/pytorchpipe/blob/f2d5ef0b96dafaae38a2f1d490f85cc05fc76dde/ptp/workers/worker.py#L297
Besides, I was passing the seed to DataLoader workers:
https://github.com/IBM/pytorchpipe/blob/f2d5ef0b96dafaae38a2f1d490f85cc05fc76dde/ptp/application/task_manager.py#L62

I wasn't setting cudnn though...

Note: I hear what Boris and Oleksii is saying about "reproducibility", just adding that here as food for thoughts ;)

Comment (blisc) [reply nemo/core/neural_factory.py]: I put a comment in slack. Namely numpy's random state has 2^{19937}-1 possible internal states. You can't just log a 32-bit number but make numpy deterministic. It's state is stored as a 624-long list of 32-bit numbers

File: examples/asr/jasper_an4.py (hunk 1 above)
Comment (blisc) [reply examples/asr/jasper_an4.py]: Actually I think you're right. I should turn this off for Jasper

File: examples/tts/tacotron2.py (hunk 2 above)
Comment (blisc) [reply examples/tts/tacotron2.py]: I switched from train loss to eval loss for tts test.

File: examples/asr/jasper_an4.py (hunk 1 above)
Comment (blisc) [reply examples/asr/jasper_an4.py]: I added eval_freq to our Jenkins CI job which is larger than the number of steps that this test trains for so eval should no longer happen.
I still need the eval modules so I can test WER after training.

File: tests/integration/test_tts_gradient_step.py [hunk 4]
```diff
@@ -137,17 +145,22 @@ def test_tacotron2_training(self):
...
         # Instantiate an optimizer to perform `train` action
         optimizer = nemo.backends.pytorch.actions.PtActions()
         optimizer.train(
-            [loss_t], callbacks=[callback], optimizer="sgd", optimization_params={"num_epochs": 10, "lr": 0.0003},
+            [loss_t], callbacks=[callback], optimizer="sgd", optimization_params={"max_steps": 3, "lr": 0.01}
         )
 
-    @pytest.mark.unclassified
+        # Assert that training loss went down
+        assert loss_list[-1] < loss_list[0]
```
Comment (tkornuta-nvidia) [inline tests/integration/test_tts_gradient_step.py:160]: 👍 🥇

Comment (tkornuta-nvidia) [reply tests/integration/test_tts_gradient_step.py:160]: Yap, looks like a great integration test, nice! :)

File: tests/integration/test_tts_gradient_step.py [hunk 5]
```diff
@@ -183,15 +196,19 @@ def test_waveglow_training(self):
...
         )
         # Instantiate an optimizer to perform `train` action
         optimizer = nemo.backends.pytorch.actions.PtActions()
         optimizer.train(
-            [loss_t], callbacks=[callback], optimizer="sgd", optimization_params={"num_epochs": 10, "lr": 0.0003},
+            [loss_t], callbacks=[callback], optimizer="sgd", optimization_params={"max_steps": 3, "lr": 0.01}
         )
 
+        # Assert that training loss went down
+        assert loss_list[-1] < loss_list[0]
```
Comment (tkornuta-nvidia) [inline tests/integration/test_tts_gradient_step.py:210]: The same here 🥇

File: tests/integration/test_tts_gradient_step.py [hunk 6]
```diff
@@ -273,11 +292,15 @@ def test_fastspeech(self):
...
         )
         # Instantiate an optimizer to perform `train` action
         optimizer = nemo.backends.pytorch.actions.PtActions()
         optimizer.train(
-            [loss_t], callbacks=[callback], optimizer="sgd", optimization_params={"num_epochs": 3, "lr": 0.0003},
+            [loss_t], callbacks=[callback], optimizer="sgd", optimization_params={"max_steps": 3, "lr": 0.0003}
         )
+
+        # Assert that training loss went down
+        assert loss_list[-1] < loss_list[0]
```
Comment (tkornuta-nvidia) [inline tests/integration/test_tts_gradient_step.py:306]: 👍

File: Jenkinsfile [hunk 7]
```diff
@@ -35,15 +35,15 @@ pipeline {
 
     stage('L0: Unit Tests GPU') {
       steps {
-        sh 'pytest -m unit'
+        sh 'pytest -m "unit and not skipduringci"'
```
Comment (tkornuta-nvidia) [inline Jenkinsfile:38]: I am not really sure why we need that for...

The simplest way to skip a test during CI is *not use any of the existing markers* - it test is not "unit", "integration", "system" or "unclassified", it won't be executed during CI...

Comment (blisc) [reply Jenkinsfile:38]: I should all tests should be classified. However in the case of TTS, there was an overlap in testing tacotron2 since it was both an integration test and a CI test. I think it would be beneficial to have this flag, if in the future, we want to pick which test to run.

Comment (tkornuta-nvidia) [reply Jenkinsfile:38]: ok, so let us leave it for now
